Add 'From Zero to Zarr' beginner guide to the Zarr data model (#4077)
zarr-developers/zarr-python: docs/user-guide/data_model.md

*From a NumPy array to stored bytes, chunk by chunk.*

This page is for people who are new to Zarr. You don't need to know NumPy, HDF5,
or anything about file formats. We begin with *why* Zarr exists, then build up
the *how* one idea at a time, until you understand **how Zarr stores an array**,
**why** that layout is defined by a written specification, and **how a library
turns those stored bytes back into an array you can use**.

The page comes in three parts:

- **Part I: The core idea.** The happy path, with pictures and no code.
- **Part II: Under the hood.** A few deeper sections that go *off* the happy
  path. Each one is signposted, so you can read on or skip ahead.
- **Part III: Seeing it for real.** A short hands-on section with runnable
  code that ties everything together.

But before the *how*, a word on the *why*.

---

## Why we need Zarr

Across science and industry, our instruments and simulations have become
extraordinary firehoses of numbers. A satellite streams images of the Earth; a
microscope captures gigapixel scans; a gene sequencer reads thousands of genomes;
a climate model writes out temperature and wind for every point on the globe, hour

> after hour. In each case the result has the same shape: a vast grid of numbers —

Reviewer: many people have a negative reaction to em-dashes since their proliferation by AI. It would likely be worth reducing their use in this guide via more, shorter sentences.

far more than fits in any single computer's memory, and often arriving as a
continuous stream.

That data is worth little sitting on one machine. It has to be stored somewhere
**durable** and **shareable**, so that many people — often scattered across the
world — can read and analyze it. Increasingly that somewhere is **cloud object
storage** (such as Amazon S3, Google Cloud Storage, or Azure Blob Storage): cheap,
effectively unlimited, and reachable from anywhere. But sheer size makes this
hard: nobody wants to download terabytes just to inspect one corner. What's needed
is a way to store these giant grids so a reader can efficiently and cheaply fetch
**just the piece they want**.

Zarr was built to solve exactly this — though, interestingly, it didn't begin with
the cloud. It grew out of **genomics**. Around 2015, Alistair Miles needed to
analyze arrays of genetic variation across thousands of malaria-carrying mosquitoes
(the [*Anopheles gambiae* 1000 Genomes Project](https://www.malariagen.net/)) —
arrays far too big to fit in memory. His real frustration was *speed*, and to see
why, it helps to understand two things the array formats of the day were already

> doing.

Reviewer: ```suggestion
doing: chunking and compression.
```
If I'm reading this right, it's not totally obvious what "Second" is

a big analysis that scans millions of values, that work adds up fast.

Here's where speed came in. Reading a chunk means decompressing it, so reading
*many* chunks is a pile of independent decompression jobs — exactly the kind of
work you'd want to spread across all your CPU cores at once. But the tools of the
day wouldn't let him: reading through HDF5 held Python's **global interpreter
lock** (the *GIL* — a safeguard that lets only one thread run Python code at a
time, so extra threads just waited their turn), and the other chunked format he
tried could chop an array along only its **first dimension**. But scientific arrays
usually have *several* dimensions — independent axes along which the data is
arranged. Alistair's mosquito data, for instance, was indexed by position along the
genome, by which mosquito was sampled, and by which of the two copies of each
chromosome it came from — three dimensions; a climate dataset might be indexed by
time, latitude, and longitude. His analyses kept needing pieces that cut across
these dimensions, and chunking along just one of them made that painfully slow. One

> core did all the work while the remaining cores sat idle.

Reviewer: this paragraph may loose a lot of people. Consider using diagrams to explain the concepts

Author: This paragraph folds together a few separate concepts: a read over many chunks becoming many independent decompression jobs; the GIL limiting concurrency in Python; chunking along a single dimension making cross-cutting reads slow; and parallel decompression across cores as the payoff. Which of these did you most want illustrated? I've also simplified the prose here to lighten the load, which may help on its own.

the whole array finishes in a fraction of the time. (He tells the story in his early
[Zarr blog posts](http://alimanfoo.github.io/2016/05/16/cpu-blues.html).)

Storing data in the cloud came **later**, and turned out to be a superpower:
because each chunk is simply one key/value entry (as we'll see), Zarr maps
naturally onto object storage like S3, which made it a backbone of cloud-native
science. Today Zarr is used far beyond genomics — in **Earth and climate science**
(satellite imagery, weather and climate model output — the
[Pangeo](https://www.pangeo.io/) community), **bio-imaging** (huge microscopy
volumes, via [OME-Zarr](https://ngff.openmicroscopy.org/)), **astronomy**, and
**machine learning** — anywhere people wrestle with large, multi-dimensional grids
of numbers.

Strip away the domain — mosquitoes, galaxies, hurricanes — and the object at the
center is always the same: an **array**, a big grid of numbers. So that's where
we'll begin. In Part I we'll look at what an array is, then at what happens
when one grows too big to fit in memory, and build up from there to how Zarr
stores it.

---

## Part I: The core idea

> ### A quick array refresher

Reviewer: ```suggestion
### An overview of arrays
```
Let's not assume familiarity

The thing Zarr stores is an **array**: a grid of values that all share a single
**data type** (almost always shortened to **dtype**, the term we'll use from here
on), arranged by a **shape**.

Here is a small array with **4 rows and 6 columns** of 32-bit integers (we use a
non-square shape on purpose, so "rows" and "columns" are never ambiguous):

<figure>
<table>
<tr><td>0</td><td>1</td><td>2</td><td>3</td><td>4</td><td>5</td></tr>
<tr><td>6</td><td>7</td><td>8</td><td>9</td><td>10</td><td>11</td></tr>
<tr><td>12</td><td>13</td><td>14</td><td>15</td><td>16</td><td>17</td></tr>
<tr><td>18</td><td>19</td><td>20</td><td>21</td><td>22</td><td>23</td></tr>
</table>
<figcaption><code>shape = (4, 6)</code> — 4 rows, 6 columns; <code>dtype = int32</code> — every value is a 32-bit integer.</figcaption>
</figure>

A few terms we'll use throughout:

- **dtype** — every element has the *same* type. Here it's a 32-bit integer, so
  each value takes exactly 4 bytes. A uniform type means the computer knows
  precisely how many bytes each value occupies, and where each one lives.
- **shape** — the size along each dimension. Ours is `(4, 6)`: the first number is
  rows, the second is columns.

> - **ndim** — the number of dimensions (axes). Ours is 2. Arrays can be 1-D, 2-D,

Reviewer: ```suggestion
- **ndim** — the number of dimensions (axes). Ours is 2. Arrays can be 0-D, 1-D, 2-D,
```
Scalar arrays are 0-D, right?

</table>
<figcaption><code>a[1:3, 2:5]</code> selects the shaded region (rows 1–2, columns 2–4).</figcaption>
</figure>

That `start:stop` bracket notation is Python and NumPy's; other languages and Zarr
implementations express the same idea with their own syntax (Rust's `s![1..3, 2..5]`,
for instance). The notation isn't part of Zarr — what matters here is the universal
*concept*: picking out a sub-region of the array.

### When an array outgrows memory

The catch with an in-memory array is right there in the name: it lives in memory,
and memory runs out. As we saw, real datasets are routinely **too big to fit in
RAM**, need to **outlive the program that created them**, and must be **shared** so
others can read even a single corner without copying the whole thing.

An array that large is never held in memory all at once — it's written out a piece
at a time (more on that in Part II). To make that possible, Zarr starts with
one simple idea: don't store the array as a single blob. Split it up.

### Chunking: splitting the grid into blocks

To store an array that may be enormous, Zarr first cuts the grid into a

> [regular pattern of equal-shaped blocks](https://zarr-specs.readthedocs.io/en/latest/v3/chunk-grids/regular-grid/index.html)

Reviewer: ```suggestion
[pattern of blocks](https://zarr-specs.readthedocs.io/en/latest/v3/chunk-grids/regular-grid/index.html)
```
the rectilinear chunk grid means it doesn't need to be regular or equal-shaped

Author: This links to the regular-grid spec, where chunks are equal-sized by definition; the rectilinear (non-equal) chunk grid is a separate extension, which the guide introduces later in the chunk-shape section. I've kept "regular grid of equal-sized blocks" here so the default is precise, but happy to add "(the default)" if you think it still over-claims.

the array's top-right block:

<figure>
<div style="display:grid;grid-template-columns:repeat(2, max-content);gap:1rem;justify-content:center">
<div style="text-align:center">
<table><tr><td>0</td><td>1</td><td>2</td></tr><tr><td>6</td><td>7</td><td>8</td></tr></table>
<small>chunk (0, 0)</small>
</div>
<div style="text-align:center">
<table><tr><td>3</td><td>4</td><td>5</td></tr><tr><td>9</td><td>10</td><td>11</td></tr></table>
<small>chunk (0, 1)</small>
</div>
<div style="text-align:center">
<table><tr><td>12</td><td>13</td><td>14</td></tr><tr><td>18</td><td>19</td><td>20</td></tr></table>
<small>chunk (1, 0)</small>
</div>
<div style="text-align:center">
<table><tr><td>15</td><td>16</td><td>17</td></tr><tr><td>21</td><td>22</td><td>23</td></tr></table>
<small>chunk (1, 1)</small>
</div>
</div>
<figcaption>A <code>(4, 6)</code> array with chunk shape <code>(2, 3)</code> forms a <strong>chunk grid</strong> of shape <code>(2, 2)</code>. Don't confuse the two: the <em>chunk shape</em> <code>(2, 3)</code> is the size of each block; the <em>chunk grid shape</em> <code>(2, 2)</code> is how many blocks there are along each axis.</figcaption>

> </figure>

Reviewer: The rendering of this figure seems off, with columns labelled as rows and index of chunks being wrong.

<img width="717" height="194" alt="Image" src="https://github.com/user-attachments/assets/e3717fa6-8663-4b0f-bdf4-b3184070f988" />

Chunking is the key move. Each chunk can be stored, loaded, and compressed on its
own, so a program can read just the chunks it needs — that one corner your

> colleague wanted — without touching the rest. (Starting with a chunk shape that

Reviewer 2: could we use an inline admonition for the partial-chunk callout? something like

> [!NOTE]
> If each chunk has a fixed size, how can we use chunks to represent an array that isn't evenly divided by the chunk size? See #section for the answer to that question!

not sure if note is the right admonition here

a store is simply *a mapping from keys to values* — where a **key** is a text
string and a **value** is a sequence of **bytes**. In other words, a store is
basically a dictionary: hand it a key, get back some bytes.

That abstraction is deliberately humble, because lots of things can play the role
of a store: a **directory on your disk** (keys are file paths), an **object-storage
bucket** like Amazon S3 (keys are object names), a **ZIP file**, or even plain
memory. Zarr treats them all the same way.

Each cell of the **chunk grid** becomes one value in the store, under a key built
from the cell's position. So the 2×2 chunk grid on the left flattens into four
key→bytes entries on the right:

```mermaid
flowchart LR
  subgraph grid ["chunk grid (2&times;2)"]
    direction TB
    subgraph gr0 ["row 0"]
      direction LR
      G00["chunk (0,0)"]
      G01["chunk (0,1)"]
    end
    subgraph gr1 ["row 1"]
      direction LR
      G10["chunk (1,0)"]
      G11["chunk (1,1)"]
    end
  end
  subgraph store ["store (key &rarr; bytes)"]
    direction TB
    K00["c/0/0"]
    K01["c/0/1"]
    K10["c/1/0"]
    K11["c/1/1"]
  end
  G00 --> K00
  G01 --> K01
  G10 --> K10
  G11 --> K11
```

> Where does a key like `c/0/1` come from? It's built by a simple, fixed rule (the

Reviewer 2: "fixed rule" locally implies that arrays have 1 chunk key encoding. maybe we can rephrase to make it clear that there's a rule defined by a particular field in array metadata.

the array has **5 rows** and we keep a chunk height of **2**?

The chunk grid simply rounds up: 5 rows with a chunk height of 2 gives row-chunks
covering rows 0–1, 2–3, and 4. That last row-chunk only has *one* real row, but —
per the [spec](https://zarr-specs.readthedocs.io/en/latest/v3/chunk-grids/regular-grid/index.html) —
**border chunks are always stored at full size**. The cells beyond the array's edge
are unused; the spec *recommends* writing the **fill value** into them.

<div style="display:flex;flex-wrap:wrap;gap:1rem">
<figure>
<table><tr><td>24</td><td>25</td><td>26</td></tr><tr><td style="opacity:0.45">fill</td><td style="opacity:0.45">fill</td><td style="opacity:0.45">fill</td></tr></table>
<figcaption>bottom chunk (2, 0)</figcaption>
</figure>
<figure>
<table><tr><td>27</td><td>28</td><td>29</td></tr><tr><td style="opacity:0.45">fill</td><td style="opacity:0.45">fill</td><td style="opacity:0.45">fill</td></tr></table>
<figcaption>bottom chunk (2, 1)</figcaption>
</figure>
</div>

So a 5×6 array chunked at `(2, 3)` quietly stores a row of "phantom" cells holding
the fill value. It's harmless, but it's a small waste — and a good reason to pick a

> chunk shape that fits your array's real shape reasonably well. (For practical

Reviewer: ```suggestion
chunk shape that fits your array's real shape reasonably well and lean on the
[rectilinear chunk grid extension](https://github.com/zarr-developers/zarr-extensions/tree/main/chunk-grids/rectilinear) when needed. (For practical
```

guidance on choosing chunk shapes, see [Performance](performance.md).)

### Going deeper: codecs (how values become bytes)

One more thing happens inside each chunk. The bytes stored under `c/0/1` aren't
necessarily the raw values — they're produced by a **codec pipeline**, a small
ordered assembly line recorded in the metadata. The
[specification](https://zarr-specs.readthedocs.io/en/latest/v3/core/index.html#codecs)
defines three kinds of codec, applied in this order:

> 1. **array → array** codecs (optional, any number) — rearrange the values; e.g. a

Reviewer: ```suggestion
1. **array → array** codecs (optional, any number) — rearrange or change the values; e.g. a
```
I believe this change is more accurate, but would appreciate if @d-v-b confirms

Author: Updated to "rearrange or transform the values" (using "transform" rather than "change"). @d-v-b, does that read accurately to you for the array→array category?

   changes their order.
2. **array → bytes** codec (exactly one, always required) — turns the array of
   values into a flat sequence of bytes. By default
   ([the `bytes` codec](https://zarr-specs.readthedocs.io/en/latest/v3/codecs/bytes/index.html))
   it writes them in lexicographical order, which the spec notes *is* C / row-major
   order.
3. **bytes → bytes** codecs (optional, any number) — transform the bytes; e.g.
   **compression** to shrink them, or a checksum to detect corruption.

Because the metadata records the exact pipeline, any spec-compliant reader knows
precisely how to run it in reverse and **decode** a chunk back into values.
Per-chunk compression is a big part of why Zarr can store enormous arrays
efficiently while staying readable everywhere.

### Going deeper: sharding (when there are too many chunks)

So far, **each chunk has been its own store object** — one key, one value. That's
simple, but it has a limit: small chunks in a very large array produce a *huge*
number of chunks, and therefore a huge number of files or objects. The spec notes
this is exactly where file systems (block sizes, inode limits) and object stores

> (which dislike millions of tiny objects) start to struggle.

Reviewer: I think the more prevalent limitation on object stores is the cost model, where the cost of operations often scales with the number of objects

prefix. So nothing new is needed to support hierarchies: the same simple rules —
keys, bytes, and metadata — scale from a single array up to a richly structured
dataset, and they map just as naturally onto a flat object store (where keys never
were folders) as onto a directory on disk. See [Groups](groups.md) to work with
them.

### Going deeper: more than two dimensions

We've stayed in 2-D to keep the pictures simple, but **nothing about Zarr is limited
to two dimensions** — the whole model scales to any number of axes, with one more
index at each step. An *N*-dimensional array's `shape` and chunk shape each gain an
axis, its chunk grid becomes *N*-dimensional, and each chunk key carries *N*
indices. The store, the metadata, and the codecs are all unchanged.

This matters because real data is *usually* more than 2-D. A short video clip is
3-D (frames × height × width); an RGB image is 3-D (height × width × colour
channel); a microscope scan or a CT volume is a stack of 2-D slices; and a climate
field recorded over time is (time × latitude × longitude). Many datasets have four
or more axes.

To see the generalisation concretely, picture a 3-D array as a **stack of 2-D

> arrays**. Here are two copies of our 4×6 grid stacked into a `(2, 4, 6)` array —

Reviewer: ```suggestion
arrays**. Here are two versions of our 4×6 grid stacked into a `(2, 4, 6)` array —
```

- The **chunk grid** is now 3-D: shape `(2, 2, 2)` — two layers × two chunk-rows ×
  two chunk-columns = **eight** chunks.
- **Keys** gain an index. The chunk at grid position `(layer, row, col) = (1, 0, 1)`
  is stored under the key `c/1/0/1`.
- **Slicing** gains an index too: `a[0]` selects the whole first layer, and
  `a[0, 1:3, 2:5]` selects a region within it.

Adding dimensions changes the *numbers* — more entries in the shape, more indices in
each key — but not the *model*. And these higher-dimensional arrays are often
exactly the ones too big for memory, which brings us to the last piece.

### Going deeper: working with data bigger than memory

Back to the question from Part I: how do you handle an array that's too big
for RAM? The answer falls out of everything above. Creating a Zarr array doesn't
allocate the whole thing — it just writes the metadata and prepares an (empty)
store. You then fill the array **a region at a time**, and each write only needs
*that region* in memory:

- read or generate one block of data (say, a few chunks' worth),
- write it to the corresponding slice of the array,
- discard it, and move on to the next block.

> Because only one block is ever in memory, the array on disk can be far larger than

Reviewer: ```suggestion
Because the minimum amount of data ever needed in memory to be useful is a single block, the array on disk can be far larger than
```

Author: Reworded to: "Because you only need to hold one block in memory at a time, the array on disk can be far larger than your RAM." That keeps your accuracy point (you never have to hold the whole array, only a block at a time) while staying a little more concise than "the minimum amount of data ever needed in memory to be useful is a single block."
